File: JavaScript/684-Redundant Connection.js
/**
 * @param {number[][]} edges
 * @return {number[]}
 */
var findRedundantConnection = function(edges) {
    const parents = new Map();
    const n = edges.length;
    
    for (let i=1; i<=n; i++){
        parents.set(i, i);
    }
    
    let find = (node) =>{
        let path = [];
        while (node !== parents.get(node)){
            path.push(node);
            node = parents.get(node);
        }
        for (let n of path){
            parents.set(n, node);
        }
        return node;
    }
    
    for (let edge of edges){
        let parent1 = find(edge[0]), parent2 = find(edge[1]);
        if (parent1 === parent2){
            return edge;
        }
        parents.set(parent2, parent1);   // union
    }
};

File: C++/684-Redundant Connection.cpp
struct UnionFind {
    vector<int> parent, rank;
    explicit UnionFind(int n) : parent(vector<int>(n)), rank(vector<int>(n, 1)) {
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int node) {
        if(node == parent[node]) return node;
        return parent[node] = find(parent[node]);
    }
    bool unite(int node1, int node2) {
        int p1 = find(node1);
        int p2 = find(node2);
        if(p1 == p2) return false; // Same parent
        if(rank[p1] >= rank[p2]) {
            // Append p2 under p1
            parent[p2] = p1;
            rank[p1] += rank[p2];
        }else {
            parent[p1] = p2;
            rank[p2] += rank[p1];
        }
        return true;
    }
};

class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int N = edges.size();
        vector<int> res = {-1, -1};
        UnionFind uf(N+1);

        for (auto &e:edges) {
            int u = e[0], v = e[1];
            if (uf.find(u) != uf.find(v)) {
                uf.unite(u, v);
            } else {
                res = {u, v};
            }
        }

        return res;
    }
};